# Predicting Housing Prices in Ames, Iowa
## Part 2 - Linear Regression
### Lisa Hwang
#### 1/17/2020

We've receieved a dataset, the Ames Housing Dataset, containing variables that were used by the Ames, Iowa, Assessor's Office to assess residential properties sold between 2006 to 2010. A total of 82 variables were assessed, and more information on them was found at this link: http://jse.amstat.org/v19n3/decock/DataDocumentation.txt.

Our goal here is to build a model which is best for predicting house sale prices. We have two csv files: train and test. I cleaned up the train dataset in the previous notebook (train_clean.csv) and will be using it here.

To recap, I plan to use these variables in my first model: 
 1. Overall Qual: overall quality
 2. Gr Liv Area: above ground living area
 3. Garage Area: garage area
 4. Total Bsmt SF: total basement square feet
 5. 1st Flr SF: first floor square feet
 6. Year Built: year built
 7. Year Remod/Add: year remodeled or addition built
 8. Condition 2_Feedr: proximity to condition (adjacent to feeder street)
 9. Condition 2_Norm: proximity to condition (normal)
 10. Condition 2_PosA: proximity to condition (adjacent to postive off-site feature such as park, greenbelt)
 11. Kitchen Qual_Fa: kitchen quality (fair)
 12. Kitchen Qual_Gd: kitchen quality (good)
 13. Kitchen Qual_TA: kitchen quality (typical/average)

### Importing libraries

In [1]:
import pandas as pd

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score

### Importing the data

In [2]:
train = pd.read_csv('Datasets/train_clean.csv')
test = pd.read_csv('Datasets/test.csv')

### Reviewing the data

In [3]:
# So I can see all of the columns in the dataframe
pd.set_option('display.max_columns', 100)

In [4]:
# Checking that the import worked
train.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0       IR1          Lvl    AllPub    CulDSac        Gtl       Sawyer   
1       IR1          Lvl    AllPub    CulDSac        Gtl      SawyerW   
2       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
3       Reg          Lvl    AllPub     Inside        Gtl       Timber   
4       IR1          Lvl    AllPub     Inside        Gtl      SawyerW   

  Condition 1 Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0        RRAe      1Fam      2Story             6             8        1976   
1        Norm      1Fam      2Story             7             5        1996   
2        Norm      1Fam      1Story             5             7        1953   
3        Norm      1Fam      2Story             5             5        2006   
4        Norm      1Fam      1.5Fin             6             8        1900   

   Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type  \
0            2005      Gable   CompShg      HdBoard      Plywood      BrkFace   
1            1997      Gable   CompShg      VinylSd      VinylSd      BrkFace   
2            2007      Gable   CompShg      VinylSd      VinylSd         None   
3            2007      Gable   CompShg      VinylSd      VinylSd         None   
4            1993      Gable   CompShg      Wd Sdng      Plywood         None   

   Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual Bsmt Cond  \
0         289.0         Gd         TA     CBlock        TA        TA   
1         132.0         Gd         TA      PConc        Gd        TA   
2           0.0         TA         Gd     CBlock        TA        TA   
3           0.0         TA         TA      PConc        Gd        TA   
4           0.0         TA         TA      PConc        Fa        Gd   

  Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  \
0            No            GLQ         533.0            Unf           0.0   
1            No            GLQ         637.0            Unf           0.0   
2            No            GLQ         731.0            Unf           0.0   
3            No            Unf           0.0            Unf           0.0   
4            No            Unf           0.0            Unf           0.0   

   Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air Electrical  \
0        192.0          725.0    GasA         Ex           Y      SBrkr   
1        276.0          913.0    GasA         Ex           Y      SBrkr   
2        326.0         1057.0    GasA         TA           Y      SBrkr   
3        384.0          384.0    GasA         Gd           Y      SBrkr   
4        676.0          676.0    GasA         TA           Y      SBrkr   

   1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  \
0         725         754                0         1479             0.0   
1         913        1209                0         2122             1.0   
2        1057           0                0         1057             1.0   
3         744         700                0         1444             0.0   
4         831         614                0         1445             0.0   

   Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  Kitchen AbvGr  \
0             0.0          2          1              3              1   
1             0.0          2          1              4              1   
2             0.0          1          0              3              1   
3             0.0         

In [5]:
test.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0  Grvl       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
1   NaN       IR1          Lvl    AllPub     Inside        Gtl       Sawyer   
2   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   
3   NaN       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl        NAmes   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        Norm        Norm    2fmCon      2Story             6             8   
1        Norm        Norm    Duplex      1Story             5             4   
2        Norm        Norm      1Fam      2Story             7             5   
3        Norm        Norm      1Fam      1Story             5             6   
4        Norm        Norm      1Fam      1Story             6             5   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1910            1950      Gable   CompShg      AsbShng      AsbShng   
1        1977            1977      Gable   CompShg      Plywood      Plywood   
2        2006            2006      Gable   CompShg      VinylSd      VinylSd   
3        1923            2006      Gable   CompShg      Wd Sdng      Wd Sdng   
4        1963            1963      Gable   CompShg      Plywood      Plywood   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0         None           0.0         TA         Fa      Stone        Fa   
1         None           0.0         TA         TA     CBlock        Gd   
2         None           0.0         Gd         TA      PConc        Gd   
3         None           0.0         Gd         TA     CBlock        TA   
4      BrkFace         247.0         TA         TA     CBlock        Gd   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        TA            No            Unf             0            Unf   
1        TA            No            Unf             0            Unf   
2        Gd            Av            GLQ           554            Unf   
3        TA            No            Unf             0            Unf   
4        TA            No            BLQ           609            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0             0         1020           1020    GasA         Gd           N   
1             0         1967           1967    GasA         TA           Y   
2             0          100            654    GasA         Ex           Y   
3             0          968            968    GasA         TA           Y   
4             0          785           1394    GasA         Gd           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      FuseP         908        1020                0         1928   
1      SBrkr        1967           0                0         1967   
2      SBrkr         664         832                0         1496   
3      SBrkr         968           0                0          968   
4      SBrkr        1394           0                0         1394   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0               0               0          2          0              4   
1               0               0          2          0              6   
2               1               0          2          1              3   
3               0               0        

I'll check if there are any nulls in the test dataset. There are columns with missing values, but none happen to be in the variables I'll use in my model.

In [6]:
test.isnull().sum().sort_values(ascending=False).head(30)

Pool QC           874
Misc Feature      837
Alley             820
Fence             706
Fireplace Qu      422
Lot Frontage      160
Garage Cond        45
Garage Qual        45
Garage Yr Blt      45
Garage Finish      45
Garage Type        44
Bsmt Exposure      25
BsmtFin Type 1     25
Bsmt Qual          25
BsmtFin Type 2     25
Bsmt Cond          25
Mas Vnr Area        1
Mas Vnr Type        1
Electrical          1
Overall Cond        0
Exter Qual          0
Exter Cond          0
Foundation          0
Exterior 2nd        0
Exterior 1st        0
Roof Matl           0
Roof Style          0
Year Remod/Add      0
Year Built          0
Sale Type           0
dtype: int64

The train dataset has 2047 rows and 86 columns. The test dataset has 878 rows and 80 columns. It is obviously missing SalePrice but I will also need to add the additional dummy variables for Condition 2 and Kitchen Quality.

In [7]:
train.shape

(2047, 85)

In [8]:
test.shape

(878, 80)

In [9]:
# Looking at the value counts for Condition 2 in test
test['Condition 2'].value_counts()

Norm     875
Feedr      2
PosA       1
Name: Condition 2, dtype: int64

For Condition 2, there are only 3 different values in the test data. This has already been taken into account in the train_clean.csv.

In [10]:
# Looking at the value counts for Kitchen Qual in test
test['Kitchen Qual'].value_counts()

TA    447
Gd    354
Ex     53
Fa     23
Po      1
Name: Kitchen Qual, dtype: int64

In [11]:
# Adding the dummy columns to test
test = pd.get_dummies(test, columns=['Condition 2', 'Kitchen Qual'])

In [12]:
# Making sure the dummies were added
test.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0  Grvl       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
1   NaN       IR1          Lvl    AllPub     Inside        Gtl       Sawyer   
2   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   
3   NaN       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl        NAmes   

  Condition 1 Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0        Norm    2fmCon      2Story             6             8        1910   
1        Norm    Duplex      1Story             5             4        1977   
2        Norm      1Fam      2Story             7             5        2006   
3        Norm      1Fam      1Story             5             6        1923   
4        Norm      1Fam      1Story             6             5        1963   

   Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type  \
0            1950      Gable   CompShg      AsbShng      AsbShng         None   
1            1977      Gable   CompShg      Plywood      Plywood         None   
2            2006      Gable   CompShg      VinylSd      VinylSd         None   
3            2006      Gable   CompShg      Wd Sdng      Wd Sdng         None   
4            1963      Gable   CompShg      Plywood      Plywood      BrkFace   

   Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual Bsmt Cond  \
0           0.0         TA         Fa      Stone        Fa        TA   
1           0.0         TA         TA     CBlock        Gd        TA   
2           0.0         Gd         TA      PConc        Gd        Gd   
3           0.0         Gd         TA     CBlock        TA        TA   
4         247.0         TA         TA     CBlock        Gd        TA   

  Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  \
0            No            Unf             0            Unf             0   
1            No            Unf             0            Unf             0   
2            Av            GLQ           554            Unf             0   
3            No            Unf             0            Unf             0   
4            No            BLQ           609            Unf             0   

   Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air Electrical  \
0         1020           1020    GasA         Gd           N      FuseP   
1         1967           1967    GasA         TA           Y      SBrkr   
2          100            654    GasA         Ex           Y      SBrkr   
3          968            968    GasA         TA           Y      SBrkr   
4          785           1394    GasA         Gd           Y      SBrkr   

   1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  \
0         908        1020                0         1928               0   
1        1967           0                0         1967               0   
2         664         832                0         1496               1   
3         968           0                0          968               0   
4        1394           0                0         1394               1   

   Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  Kitchen AbvGr  \
0               0          2          0              4              2   
1               0          2          0              6              2   
2               0          2          1              3              1   
3               0    

I need to drop Kitchen Qual_Ex and Kitchen Qual_Po since they are not in the train dataset.

In [13]:
test.drop(columns=['Kitchen Qual_Ex', 'Kitchen Qual_Po'], inplace=True)

In [14]:
# Making sure the drop worked
test.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0  Grvl       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
1   NaN       IR1          Lvl    AllPub     Inside        Gtl       Sawyer   
2   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   
3   NaN       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl        NAmes   

  Condition 1 Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0        Norm    2fmCon      2Story             6             8        1910   
1        Norm    Duplex      1Story             5             4        1977   
2        Norm      1Fam      2Story             7             5        2006   
3        Norm      1Fam      1Story             5             6        1923   
4        Norm      1Fam      1Story             6             5        1963   

   Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type  \
0            1950      Gable   CompShg      AsbShng      AsbShng         None   
1            1977      Gable   CompShg      Plywood      Plywood         None   
2            2006      Gable   CompShg      VinylSd      VinylSd         None   
3            2006      Gable   CompShg      Wd Sdng      Wd Sdng         None   
4            1963      Gable   CompShg      Plywood      Plywood      BrkFace   

   Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual Bsmt Cond  \
0           0.0         TA         Fa      Stone        Fa        TA   
1           0.0         TA         TA     CBlock        Gd        TA   
2           0.0         Gd         TA      PConc        Gd        Gd   
3           0.0         Gd         TA     CBlock        TA        TA   
4         247.0         TA         TA     CBlock        Gd        TA   

  Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  \
0            No            Unf             0            Unf             0   
1            No            Unf             0            Unf             0   
2            Av            GLQ           554            Unf             0   
3            No            Unf             0            Unf             0   
4            No            BLQ           609            Unf             0   

   Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air Electrical  \
0         1020           1020    GasA         Gd           N      FuseP   
1         1967           1967    GasA         TA           Y      SBrkr   
2          100            654    GasA         Ex           Y      SBrkr   
3          968            968    GasA         TA           Y      SBrkr   
4          785           1394    GasA         Gd           Y      SBrkr   

   1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  \
0         908        1020                0         1928               0   
1        1967           0                0         1967               0   
2         664         832                0         1496               1   
3         968           0                0          968               0   
4        1394           0                0         1394               1   

   Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  Kitchen AbvGr  \
0               0          2          0              4              2   
1               0          2          0              6              2   
2               0          2          1              3              1   
3               0    

### Train-test split

Now I will perform a train-test split so that we can randomly divide up our train dataset into a train dataset and a test dataset. We'll use the standard 75%-25% split.

In [15]:
X = train[['Overall Qual', 
          'Gr Liv Area', 
          'Garage Area', 
          'Total Bsmt SF', 
          '1st Flr SF', 
          'Year Built', 
          'Year Remod/Add', 
          'Condition 2_Feedr', 
          'Condition 2_Norm', 
          'Condition 2_PosA', 
          'Kitchen Qual_Fa', 
          'Kitchen Qual_Gd', 
          'Kitchen Qual_TA']]
y = train['SalePrice']

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

### Modeling

Here is where I will create my linear regression model.

In [17]:
lr = LinearRegression()

In [18]:
# Generating cross validation scores
cross_val_score(lr, X_train, y_train, cv=5)

array([0.8509308 , 0.84056879, 0.85040171, 0.85963686, 0.84803333])

In [19]:
cross_val_score(lr, X_train, y_train, cv=5).mean()

0.8499142986382526

The average cross validation score is 0.85, and the five cross validation scores are all close to that value. This suggests that the model isn't being distorted by outliers, if any exist, and I'm getting consistent results with each fold.

In [20]:
# Fitting the model
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [21]:
# The model's Rsquared value on the train data
lr.score(X_train, y_train)

0.8564087829797944

In [22]:
# The model's Rsquared value on the test data
lr.score(X_test, y_test)

0.8581404213261252

The model's cross validation score, train score, and test score are all similar and relatively high. This is a indication that I have a strong model!

### Summary
The linear regression model with the variables below performed well at predicting sale prices, obtaining an Rsquared score of 0.8564 on the train data and 0.8581 on the test data. 
 - Condition 2_Feedr: proximity to condition (adjacent to feeder street)
 - Condition 2_Norm: proximity to condition (normal)
 - Condition 2_PosA: proximity to condition (adjacent to postive off-site feature such as park, greenbelt)
 - Kitchen Qual_Ex: kitchen quality (excellent)
 - Kitchen Qual_Fa: kitchen quality (fair)
 - Kitchen Qual_Gd: kitchen quality (good)
 - Kitchen Qual_TA: kitchen quality (typical/average)

### Continued in Notebook 3